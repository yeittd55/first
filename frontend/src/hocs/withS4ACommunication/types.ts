export type IMenuItem =
  | {
      type: "text"
      label: string
      key: string
    }
  | {
      type: "separator"
    }

export type IHostToGuestMessage = {
  stCommVersion: number
} & (
  | {
      type: "SET_MENU_ITEMS"
      items: IMenuItem[]
    }
  | {
      type: "UPDATE_FROM_QUERY_PARAMS"
      queryParams: string
    }
)

export type IGuestToHostMessage =
  | {
      type: "GUEST_READY"
    }
  | {
      type: "MENU_ITEM_CALLBACK"
      key: string
    }
  | {
      type: "SET_PAGE_TITLE"
      title: string
    }
  | {
      type: "SET_PAGE_FAVICON"
      favicon: string
    }
  | {
      type: "SET_QUERY_PARAM"
      queryParams: string
    }

export type VersionedMessage<Message> = {
  stCommVersion: number
} & Message
