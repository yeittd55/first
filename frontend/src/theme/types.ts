import main from "./mainTheme"

export type Theme = typeof main
type IconSizes = typeof main.iconSizes
type ThemeSpacings = typeof main.spacing
type ThemeColors = typeof main.colors

export type IconSize = keyof IconSizes
export type ThemeColor = keyof ThemeColors
export type ThemeSpacing = keyof ThemeSpacings
