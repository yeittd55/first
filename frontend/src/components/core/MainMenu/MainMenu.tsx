/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactElement, memo, forwardRef, MouseEvent } from "react"
import { StatefulPopover, PLACEMENT } from "baseui/popover"
import { StatefulMenu } from "baseui/menu"
import Button, { Kind } from "components/shared/Button"
import { Menu } from "@emotion-icons/open-iconic"

import Icon from "components/shared/Icon"
import {
  IMenuItem,
  IGuestToHostMessage,
} from "hocs/withS4ACommunication/types"

import { IDeployParams } from "autogen/proto"
import {
  BUG_URL,
  COMMUNITY_URL,
  DEPLOY_URL,
  ONLINE_DOCS_URL,
  STREAMLIT_SHARE_URL,
  TEAMS_URL,
} from "urls"
import {
  StyledMenuItem,
  StyledMenuDivider,
  StyledMenuItemLabel,
  StyledMenuItemShortcut,
  StyledRecordingIndicator,
} from "./styled-components"

const SCREENCAST_LABEL: { [s: string]: string } = {
  COUNTDOWN: "Cancel screencast",
  RECORDING: "Stop recording",
}

export interface Props {
  /** True if report sharing is properly configured and enabled. */
  sharingEnabled: boolean

  /** True if we're connected to the Streamlit server. */
  isServerConnected: boolean

  /** Rerun the report. */
  quickRerunCallback: () => void

  /** Clear the cache. */
  clearCacheCallback: () => void

  /** Show the screen recording dialog. */
  screencastCallback: () => void

  /** Share the report to S3. */
  shareCallback: () => void

  /** Show the Settings dialog. */
  settingsCallback: () => void

  /** Show the About dialog. */
  aboutCallback: () => void

  screenCastState: string

  s4aMenuItems: IMenuItem[]

  sendS4AMessage: (message: IGuestToHostMessage) => void

  deployParams?: IDeployParams | null
}

const getOpenInWindowCallback = (url: string) => (): void => {
  window.open(url, "_blank")
}

const getDeployAppUrl = (
  deployParams: IDeployParams | null | undefined
): (() => void) => {
  // If the app was run inside a GitHub repo, autofill for a one-click deploy.
  // E.g.: https://share.streamlit.io/deploy?repository=melon&branch=develop&mainModule=streamlit_app.py
  if (deployParams) {
    const deployUrl = new URL(DEPLOY_URL)

    deployUrl.searchParams.set("repository", deployParams.repository || "")
    deployUrl.searchParams.set("branch", deployParams.branch || "")
    deployUrl.searchParams.set("mainModule", deployParams.module || "")

    return getOpenInWindowCallback(deployUrl.toString())
  }

  // Otherwise, just direct them to the Streamlit Share page.
  return getOpenInWindowCallback(STREAMLIT_SHARE_URL)
}

const isLocalhost = (): boolean => {
  return (
    window.location.hostname === "localhost" ||
    window.location.hostname === "127.0.0.1"
  )
}

export interface MenuListItemProps {
  item: any
  "aria-selected": boolean
  onClick: (e: MouseEvent<HTMLLIElement>) => void
  onMouseEnter: (e: MouseEvent<HTMLLIElement>) => void
  $disabled: boolean
  $isHighlighted: boolean
}

// BaseWeb provides a very basic list item (or option) for its dropdown
// menus. We want to customize it to our liking. We want to support:
//  * Shortcuts
//  * Red coloring for the stop recording
//  * Dividers (There's no special MenuListItem divider, so items have
//    a hasDividerAbove property to add the border properly.
// Unfortunately, because we are overriding the component, we need to
// implement some of the built in-features, namely:
//  * A11y for selected and disabled
//  * $disabled field (BaseWeb does not use CSS :disabled here)
//  * $isHighlighted field (BaseWeb does not use CSS :hover here)
//  * creating a forward ref to add properties to the DOM element.
const MenuListItem = forwardRef<HTMLLIElement, MenuListItemProps>(
  (
    {
      item,
      "aria-selected": ariaSelected,
      onClick,
      onMouseEnter,
      $disabled,
      $isHighlighted,
    },
    ref
  ) => {
    const { label, shortcut, hasDividerAbove } = item
    const menuItemProps = {
      isDisabled: $disabled,
      isHighlighted: $isHighlighted,
      isRecording: Boolean(item.stopRecordingIndicator),
    }
    const interactiveProps = $disabled
      ? {}
      : {
          onClick,
          onMouseEnter,
        }

    return (
      <>
        {hasDividerAbove && <StyledMenuDivider />}
        <StyledMenuItem
          ref={ref}
          role="option"
          aria-selected={ariaSelected}
          aria-disabled={$disabled}
          {...menuItemProps}
          {...interactiveProps}
        >
          <StyledMenuItemLabel {...menuItemProps}>{label}</StyledMenuItemLabel>
          {shortcut && (
            <StyledMenuItemShortcut {...menuItemProps}>
              {shortcut}
            </StyledMenuItemShortcut>
          )}
        </StyledMenuItem>
      </>
    )
  }
)

function MainMenu(props: Props): ReactElement {
  const isServerDisconnected = !props.isServerConnected

  const coreMenuOptions = {
    DIVIDER: { isDivider: true },
    rerun: {
      disabled: isServerDisconnected,
      onClick: props.quickRerunCallback,
      label: "Rerun",
      shortcut: "r",
    },
    clearCache: {
      disabled: isServerDisconnected,
      onClick: props.clearCacheCallback,
      label: "Clear cache",
      shortcut: "c",
    },
    recordScreencast: {
      onClick: props.screencastCallback,
      label: SCREENCAST_LABEL[props.screenCastState] || "Record a screencast",
      shortcut: SCREENCAST_LABEL[props.screenCastState] ? "esc" : "",
      stopRecordingIndicator: Boolean(SCREENCAST_LABEL[props.screenCastState]),
    },
    deployApp: {
      onClick: getDeployAppUrl(props.deployParams),
      label: "Deploy this app",
    },
    saveSnapshot: {
      disabled: isServerDisconnected,
      onClick: props.shareCallback,
      label: "Save a snapshot",
    },
    documentation: {
      onClick: getOpenInWindowCallback(ONLINE_DOCS_URL),
      label: "Documentation",
    },
    community: {
      onClick: getOpenInWindowCallback(COMMUNITY_URL),
      label: "Ask a question",
    },
    report: {
      onClick: getOpenInWindowCallback(BUG_URL),
      label: "Report a bug",
    },
    s4t: {
      onClick: getOpenInWindowCallback(TEAMS_URL),
      label: "Streamlit for Teams",
    },
    settings: { onClick: props.settingsCallback, label: "Settings" },
    about: { onClick: props.aboutCallback, label: "About" },
  }

  const S4AMenuOptions = props.s4aMenuItems.map(item => {
    if (item.type === "separator") {
      return coreMenuOptions.DIVIDER
    }

    return {
      onClick: () =>
        props.sendS4AMessage({
          type: "MENU_ITEM_CALLBACK",
          key: item.key,
        }),
      label: item.label,
    }
  }, [] as any[])

  const shouldShowS4AMenu = !!S4AMenuOptions.length

  const showDeploy = isLocalhost() && !shouldShowS4AMenu
  const showSnapshot = !shouldShowS4AMenu && props.sharingEnabled
  const showClearCache = !shouldShowS4AMenu
  const preferredMenuOrder: any[] = [
    coreMenuOptions.rerun,
    showClearCache && coreMenuOptions.clearCache,
    shouldShowS4AMenu && coreMenuOptions.settings,
    coreMenuOptions.DIVIDER,
    showDeploy && coreMenuOptions.deployApp,
    showSnapshot && coreMenuOptions.saveSnapshot,
    coreMenuOptions.recordScreencast,
    ...(shouldShowS4AMenu
      ? S4AMenuOptions
      : [
          coreMenuOptions.DIVIDER,
          coreMenuOptions.documentation,
          coreMenuOptions.community,
          coreMenuOptions.report,
          coreMenuOptions.DIVIDER,
          coreMenuOptions.s4t,
          coreMenuOptions.settings,
          coreMenuOptions.about,
        ]),
  ]

  // Remove empty entries, and add dividers into menu options as needed.
  const menuOptions: any[] = []
  let lastMenuItem = null
  for (const menuItem of preferredMenuOrder) {
    if (menuItem) {
      if (menuItem !== coreMenuOptions.DIVIDER) {
        if (lastMenuItem === coreMenuOptions.DIVIDER) {
          menuOptions.push({ ...menuItem, hasDividerAbove: true })
        } else {
          menuOptions.push(menuItem)
        }
      }

      lastMenuItem = menuItem
    }
  }

  return (
    <StatefulPopover
      focusLock
      placement={PLACEMENT.bottomRight}
      content={({ close }) => (
        <StatefulMenu
          items={menuOptions}
          onItemSelect={({ item }) => {
            item.onClick()
            close()
          }}
          overrides={{
            Option: MenuListItem,
            List: {
              props: {
                "data-testid": "main-menu-list",
              },
              style: {
                ":focus": {
                  outline: "none",
                },
              },
            },
          }}
        />
      )}
      overrides={{
        Body: {
          props: {
            "data-testid": "main-menu-popover",
          },
        },
      }}
    >
      <span id="MainMenu">
        <Button kind={Kind.ICON}>
          <Icon content={Menu} />
        </Button>
        {props.screenCastState === "RECORDING" && <StyledRecordingIndicator />}
      </span>
    </StatefulPopover>
  )
}

export default memo(MainMenu)
