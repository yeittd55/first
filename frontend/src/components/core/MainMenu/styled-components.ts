import styled from "@emotion/styled"
import { keyframes } from "@emotion/core"
import { Keyframes } from "@emotion/serialize"
import { Theme } from "theme"

const recordingIndicatorPulse = (theme: Theme): Keyframes => keyframes`
0% {
  box-shadow: 0 0 ${theme.spacing.twoXS} ${theme.colors.red};
}
50% {
  box-shadow: 0 0 ${theme.spacing.sm} ${theme.spacing.twoXS} ${theme.colors.red};
}
100% {
  box-shadow: 0 0 ${theme.spacing.twoXS} ${theme.colors.red};
}`

export const StyledRecordingIndicator = styled.div(({ theme }) => ({
  position: "absolute",
  bottom: theme.spacing.lg,
  right: theme.spacing.sm,
  width: theme.spacing.sm,
  height: theme.spacing.sm,
  backgroundColor: "red",
  borderRadius: theme.radii.full,
  boxShadow: `0 0 ${theme.spacing.twoXS} ${theme.colors.red}`,
  animation: `${recordingIndicatorPulse(theme)} 2s linear infinite`,
}))

export const StyledMenuDivider = styled.div(({ theme }) => ({
  borderTop: `1px solid ${theme.colors.lightGray}`,
  margin: `${theme.spacing.sm} ${theme.spacing.none}`,
}))

export interface StyledMenuItemProps {
  isDisabled: boolean
  isHighlighted: boolean
  isRecording: boolean
}

export const StyledMenuItemShortcut = styled.span<StyledMenuItemProps>(
  ({ isRecording, theme }) => {
    return {
      color: isRecording ? theme.colors.red : theme.colors.gray,
      fontSize: theme.fontSizes.sm,
      marginTop: theme.spacing.twoXS,
      fontVariant: "small-caps",
      textTransform: "uppercase",
    }
  }
)

export const StyledMenuItem = styled.li<StyledMenuItemProps>(
  ({ isDisabled, isHighlighted, isRecording, theme }) => {
    const disabledStyles = isDisabled
      ? {
          backgroundColor: theme.colors.transparent,
          color: theme.colors.gray,
          cursor: "not-allowed",
        }
      : {
          "&:active": {
            backgroundColor: theme.colors.primary,
            color: theme.colors.white,
            outline: "none",
            [StyledMenuItemShortcut as any]: {
              color: theme.colors.white,
            },
          },
          "&:focus": {
            backgroundColor: theme.colors.primary,
            color: theme.colors.white,
          },
        }

    const highlightedStyles = isHighlighted && {
      backgroundColor: theme.colors.lightGray,
    }

    const recordingStyles = isRecording && {
      color: theme.colors.red,
      fontWeight: theme.fontWeights.bold,
    }

    return {
      margin: 0,
      padding: `${theme.spacing.twoXS} ${theme.spacing.twoXL}`,
      display: "flex",
      flexDirection: "row",
      alignItems: "flex-start",
      cursor: "pointer",
      ...(highlightedStyles || {}),
      ...(recordingStyles || {}),
      ...disabledStyles,
    }
  }
)

export const StyledMenuItemLabel = styled.span(({ theme }) => ({
  marginRight: theme.spacing.md,
  flexGrow: 1,
}))
