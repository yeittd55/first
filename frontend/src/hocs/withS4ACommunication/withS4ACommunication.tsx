/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ComponentType, useState, useEffect, ReactElement } from "react"
import hoistNonReactStatics from "hoist-non-react-statics"

import { CLOUD_COMM_WHITELIST } from "urls"

import {
  IMenuItem,
  IHostToGuestMessage,
  IGuestToHostMessage,
  VersionedMessage,
} from "./types"

interface State {
  queryParams: string
  items: IMenuItem[]
}

export interface S4ACommunicationHOC {
  currentState: State
  connect: () => void
  sendMessage: (message: IGuestToHostMessage) => void
}

const S4A_COMM_VERSION = 1

export function sendS4AMessage(message: IGuestToHostMessage): void {
  window.parent.postMessage(
    {
      stCommVersion: S4A_COMM_VERSION,
      ...message,
    } as VersionedMessage<IGuestToHostMessage>,
    "*"
  )
}

function withS4ACommunication(
  WrappedComponent: ComponentType<any>
): ComponentType<any> {
  function ComponentWithS4ACommunication(props: any): ReactElement {
    const [items, setItems] = useState<IMenuItem[]>([])
    const [queryParams, setQueryParams] = useState("")

    useEffect(() => {
      function receiveMessage(event: MessageEvent): void {
        let origin
        const message: VersionedMessage<IHostToGuestMessage> | any = event.data

        try {
          const url = new URL(event.origin)

          origin = url.hostname
        } catch (e) {
          origin = event.origin
        }

        if (
          !origin ||
          message.stCommVersion !== S4A_COMM_VERSION ||
          !CLOUD_COMM_WHITELIST.includes(origin)
        ) {
          return
        }

        if (message.type === "SET_MENU_ITEMS") {
          setItems(message.items)
        }

        if (message.type === "UPDATE_FROM_QUERY_PARAMS") {
          setQueryParams(message.queryParams)
        }
      }

      window.addEventListener("message", receiveMessage)

      return () => {
        window.removeEventListener("message", receiveMessage)
      }
    }, [])

    return (
      <WrappedComponent
        s4aCommunication={
          {
            currentState: {
              items,
              queryParams,
            },
            connect: () => {
              sendS4AMessage({
                type: "GUEST_READY",
              })
            },
            sendMessage: sendS4AMessage,
          } as S4ACommunicationHOC
        }
        {...props}
      />
    )
  }

  ComponentWithS4ACommunication.displayName = `withS4ACommunication(${WrappedComponent.displayName ||
    WrappedComponent.name})`

  // Static methods must be copied over
  // https://en.reactjs.org/docs/higher-order-components.html#static-methods-must-be-copied-over
  return hoistNonReactStatics(ComponentWithS4ACommunication, WrappedComponent)
}

export default withS4ACommunication
