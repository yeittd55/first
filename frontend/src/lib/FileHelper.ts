import { CancelTokenSource } from "axios"
import { isFromWindows } from "lib/utils"

export interface ExtendedFile extends File {
  id?: string
  status?: string
  errorMessage?: string
  cancelToken?: CancelTokenSource
  progress?: number
}

export enum FileStatus {
  ERROR = "ERROR",
  DELETING = "DELETING",
  READY = "READY",
  UPLOADING = "UPLOADING",
  UPLOADED = "UPLOADED",
}

export enum FileSize {
  GigaByte = "gb",
  KiloByte = "kb",
  MegaByte = "mb",
  Byte = "b",
}

// There is a shift towards displaying storage in base 10 vs base 2
// but Windows is still displaying things in base 2. This does not handle
// all cases but for simplicity general rule is to use base 2 for Windows.
export const BYTE_CONVERSION_SIZE = isFromWindows() ? 1024 : 1000
const sizeUnitSequence = [
  FileSize.GigaByte,
  FileSize.MegaByte,
  FileSize.KiloByte,
  FileSize.Byte,
]

export const getSizeDisplay = (
  size: number,
  unit: FileSize,
  rounding = 1
): string => {
  if (!unit) {
    unit = FileSize.Byte
  }

  if (rounding < 0) {
    rounding = 0
  }

  if (size < 0) {
    throw new Error("Size must be greater than or equal to 0")
  }

  const sizeIndex = sizeUnitSequence.indexOf(unit)
  const nextUnitSize = size / BYTE_CONVERSION_SIZE
  if (sizeIndex && size > BYTE_CONVERSION_SIZE / 2) {
    return getSizeDisplay(
      nextUnitSize,
      sizeUnitSequence[sizeIndex - 1],
      rounding
    )
  }
  return `${size.toFixed(rounding)}${unit.toUpperCase()}`
}

export const sizeConverter = (
  size: number,
  inputUnit: FileSize,
  outputUnit: FileSize
): number => {
  if (size < 0) {
    throw Error("Size must be 0 or greater")
  }

  const inputLevel = sizeUnitSequence.findIndex(unit => unit === inputUnit)
  const outputLevel = sizeUnitSequence.findIndex(unit => unit === outputUnit)

  if (inputLevel === -1 || outputLevel === -1) {
    // Should not ever occur
    throw Error("Unexpected byte unit provided")
  }

  if (inputLevel === outputLevel) {
    return size
  }

  const levelsBetween = Math.abs(inputLevel - outputLevel)
  const byteDifference = BYTE_CONVERSION_SIZE ** levelsBetween

  if (inputLevel > outputLevel) {
    // Going from smaller to bigger
    return size / byteDifference
  }
  // Going from bigger to smaller
  return size * byteDifference
}
