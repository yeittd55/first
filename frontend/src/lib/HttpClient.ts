/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import axios, { AxiosRequestConfig, AxiosResponse } from "axios"
import { buildHttpUri, BaseUriParts } from "lib/UriUtil"
import { getCookie } from "lib/utils"

/**
 * Base class for HTTP Clients
 */
export default class HttpClient {
  protected readonly getServerUri: () => BaseUriParts | undefined

  protected readonly csrfEnabled: boolean

  public constructor(
    getServerUri: () => BaseUriParts | undefined,
    csrfEnabled: boolean
  ) {
    this.getServerUri = getServerUri
    this.csrfEnabled = csrfEnabled
  }

  /**
   * Wrapper around axios.request to update the request config with
   * CSRF headers if client has CSRF protection enabled.
   */
  public request(
    url: string,
    params: AxiosRequestConfig
  ): Promise<AxiosResponse> {
    const serverURI = this.getServerUri()
    if (serverURI === undefined) {
      throw new Error("Cannot complete request: not connected to a server")
    }
    params.url = buildHttpUri(serverURI, url)

    if (this.csrfEnabled) {
      const xsrfCookie = getCookie("_xsrf")
      if (xsrfCookie != null) {
        params.headers = {
          "X-Xsrftoken": xsrfCookie,
          ...(params.headers || {}),
        }
        params.withCredentials = true
      }
    }

    return axios.request(params)
  }
}
