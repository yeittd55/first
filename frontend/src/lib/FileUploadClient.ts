/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { CancelToken } from "axios"
import { ExtendedFile } from "lib/FileHelper"
import HttpClient from "lib/HttpClient"
import { SessionInfo } from "lib/SessionInfo"

/**
 * Handles uploading files to the server.
 */
export class FileUploadClient extends HttpClient {
  /**
   * Upload a file to the server. It will be associated with this browser's sessionID.
   *
   * @param widgetId: the ID of the FileUploader widget that's doing the upload.
   * @param files: the files to upload.
   * @param onUploadProgress: an optional function that will be called repeatedly with progress events during the upload.
   * @param cancelToken: an optional axios CancelToken that can be used to cancel the in-progress upload.
   * @param replace: an optional boolean to indicate if the file should replace existing files associated with the widget.
   */
  public async uploadFiles(
    widgetId: string,
    files: ExtendedFile[],
    onUploadProgress?: (progressEvent: any) => void,
    cancelToken?: CancelToken,
    replace?: boolean
  ): Promise<void> {
    const form = new FormData()
    form.append("sessionId", SessionInfo.current.sessionId)
    form.append("widgetId", widgetId)
    if (replace) form.append("replace", "true")
    for (const file of files) {
      form.append(file.id || file.name, file, file.name)
    }

    await this.request("upload_file", {
      cancelToken,
      method: "POST",
      data: form,
      onUploadProgress,
    })
  }

  public async delete(widgetId: string, fileId: string): Promise<void> {
    await this.request(
      `upload_file/${SessionInfo.current.sessionId}/${widgetId}/${fileId}`,
      { method: "DELETE" }
    )
  }

  public async updateFileCount(
    widgetId: string,
    fileCount: number
  ): Promise<void> {
    await this.request(
      `upload_file/${SessionInfo.current.sessionId}/${widgetId}`,
      {
        method: "PUT",
        data: { totalFiles: fileCount },
      }
    )
  }
}
