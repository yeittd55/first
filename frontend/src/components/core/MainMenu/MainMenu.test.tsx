/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react"
import { shallow } from "lib/test_util"
import { IMenuItem } from "hocs/withS4ACommunication/types"

import MainMenu, { Props } from "./MainMenu"

const getProps = (extend?: Partial<Props>): Props => ({
  aboutCallback: jest.fn(),
  clearCacheCallback: jest.fn(),
  isServerConnected: true,
  quickRerunCallback: jest.fn(),
  s4aMenuItems: [],
  screencastCallback: jest.fn(),
  screenCastState: "",
  sendS4AMessage: jest.fn(),
  settingsCallback: jest.fn(),
  shareCallback: jest.fn(),
  sharingEnabled: false,
  ...extend,
})

describe("App", () => {
  it("renders without crashing", () => {
    const props = getProps()
    const wrapper = shallow(<MainMenu {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it("should render s4a menu items", () => {
    const items: IMenuItem[] = [
      {
        type: "text",
        label: "Share this app",
        key: "share",
      },
      {
        type: "separator",
      },
      {
        type: "text",
        label: "View app source",
        key: "source",
      },
      {
        type: "text",
        label: "Report bug with app",
        key: "support",
      },
      {
        type: "separator",
      },
      {
        type: "text",
        label: "About Streamlit for Teams",
        key: "about",
      },
    ]
    const props = getProps({
      s4aMenuItems: items,
    })
    const wrapper = shallow(<MainMenu {...props} />)
    const popoverContent = wrapper.find("StatefulPopover").prop("content")

    // @ts-ignore
    const menuWrapper = shallow(popoverContent(() => {})).dive()

    // @ts-ignore
    const menuLabels = menuWrapper
      .find("MenuStatefulContainer")
      .prop("items")
      // @ts-ignore
      .map(item => item.label)
    expect(menuLabels).toEqual([
      "Rerun",
      "Settings",
      "Record a screencast",
      "Share this app",
      "View app source",
      "Report bug with app",
      "About Streamlit for Teams",
    ])
  })

  it("should render core set of menu elements", () => {
    const props = getProps()
    const wrapper = shallow(<MainMenu {...props} />)
    const popoverContent = wrapper.find("StatefulPopover").prop("content")
    // @ts-ignore
    const menuWrapper = shallow(popoverContent(() => {})).dive()

    // @ts-ignore
    const menuLabels = menuWrapper
      .find("MenuStatefulContainer")
      .prop("items")
      // @ts-ignore
      .map(item => item.label)
    expect(menuLabels).toEqual([
      "Rerun",
      "Clear cache",
      "Deploy this app",
      "Record a screencast",
      "Documentation",
      "Ask a question",
      "Report a bug",
      "Streamlit for Teams",
      "Settings",
      "About",
    ])
  })

  it("should render deploy app menu item", () => {
    const props = getProps({ deployParams: {} })
    const wrapper = shallow(<MainMenu {...props} />)
    const popoverContent = wrapper.find("StatefulPopover").prop("content")
    // @ts-ignore
    const menuWrapper = shallow(popoverContent(() => {})).dive()

    // @ts-ignore
    const menuLabels = menuWrapper
      .find("MenuStatefulContainer")
      .prop("items")
      // @ts-ignore
      .map(item => item.label)
    expect(menuLabels).toEqual([
      "Rerun",
      "Clear cache",
      "Deploy this app",
      "Record a screencast",
      "Documentation",
      "Ask a question",
      "Report a bug",
      "Streamlit for Teams",
      "Settings",
      "About",
    ])
  })
})
