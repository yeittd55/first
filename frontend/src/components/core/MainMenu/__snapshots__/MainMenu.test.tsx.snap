// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing 1`] = `
<StatefulPopover
  accessibilityType="menu"
  content={[Function]}
  dismissOnClickOutside={true}
  dismissOnEsc={true}
  focusLock={true}
  ignoreBoundary={false}
  onMouseEnterDelay={200}
  onMouseLeaveDelay={200}
  overrides={
    Object {
      "Body": Object {
        "props": Object {
          "data-testid": "main-menu-popover",
        },
      },
    }
  }
  placement="bottomRight"
  popoverMargin={8}
  showArrow={false}
  stateReducer={[Function]}
  triggerType="click"
>
  <span
    id="MainMenu"
  >
    <Button
      kind="icon"
    >
      <Icon
        content={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          }
        }
      />
    </Button>
  </span>
</StatefulPopover>
`;
